pusherable/mixins.py

```diff
@@ -12,43 +13,141 @@
-    
-    def render_to_response(self, context, **response_kwargs):
 
-        channel = u"{model}_{pk}".format(
-            model=self.object._meta.model_name,
-            pk=self.object.pk
-        )
-        
-        data = self.__object_to_json_serializable(self.object)
-        
-        pusher = Pusher(app_id=settings.PUSHER_APP_ID,
-                        key=settings.PUSHER_KEY,
-                        secret=settings.PUSHER_SECRET)
-        pusher.trigger(
-            [channel, ],
-            self.pusher_event_name,
-            {
-                'object': data,
-                'user': self.request.user.username
-            }
-        )
-
-        return super(PusherMixin, self).render_to_response(context, **response_kwargs)
-        
-    def __object_to_json_serializable(self, object):
-        model_dict = model_to_dict(object, 
-                                   fields=self.pusher_include_model_fields, exclude=self.pusher_exclude_model_fields)
+    def _object_to_json_serializable(self, object):
+        model_dict = model_to_dict(
+            object, fields=self.pusher_include_model_fields,
+            exclude=self.pusher_exclude_model_fields)
         json_data = json.dumps(model_dict, cls=DjangoJSONEncoder)
         data = json.loads(json_data)
+
         return data
 
+    def set_pusher(self):
+        """
+        Check that pusher settings exist or raise ValidationError.
+        Create Pusher on object.
+        """
+        if hasattr(self, 'pusher') and isinstance(self.pusher, Pusher):
+            return
+
+        app_id = getattr(settings, 'PUSHER_APP_ID', None)
+        key = getattr(settings, 'PUSHER_KEY', None)
+        secret = getattr(settings, 'PUSHER_SECRET', None)
+
+        if not all([app_id, key, secret]):
+            raise ImproperlyConfigured(
+                'Pusher settings are not defined. Make sure PUSHER_APP_ID, '
+                'PUSHER_KEY and PUSHER_SECRET are set in your settings file.')
+
+        self.pusher = Pusher(
+            app_id=getattr(settings, 'PUSHER_APP_ID'),
+            key=getattr(settings, 'PUSHER_KEY'),
+            secret=getattr(settings, 'PUSHER_SECRET'))
+
+    def set_pusher_channel(self):
+        """
+        Set the pusher channel
+        <model_name>_<primary_key>
+        """
+        self.channel = '{model}_{pk}'.format(
+            model=self.object._meta.model_name,
+            pk=self.object.pk)
+
+    def get_pusher_payload(self, data):
+        """
+        Method responsible for building payload data for pusher.
+        Override this method to customize the data sent to pusher.
+        Method should return a dict.
+        """
+        user = self.request.user
+        username = user.username
+
+        if user.is_anonymous():
+            username = 'Anonymous User'
+
+        return {
+            'object': data,
+            'user': username
+        }
+
+    def send_pusher_notification(self, event):
+        self.set_pusher()
```

Comment:
The functional programmer in me is disgusted by this and the next line. Since you never call `set_pusher()` anywhere but here, why not just do `self.channel = self.set_pusher()` and *not* modify outside state inside of a function?

Same for the next line.